src/commands/cmd_query.c

```diff
@@ -166,19 +141,29 @@ int MGraph_Query(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
         return REDISMODULE_OK;
     }
 
-    bool readonly = AST_ReadOnly(ast);
-
-    // Construct concurent query context.
-    RedisModuleBlockedClient *bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
-
-    QueryContext *context = _queryContext_New(bc, ast, argv[1]);
-
-    context->tic[0] = tic[0];
-    context->tic[1] = tic[1];    
-
-    thpool_add_work(_thpool, _MGraph_Query, context);
+    /* Determin query execution context
+     * queries issued within a LUA script or multi exec block must
+     * run on Redis main thread, others can run on different threads. */
+    CommandCtx *context;
+    int flags = RedisModule_GetContextFlags(ctx);
+    if (flags & (REDISMODULE_CTX_FLAGS_MULTI | REDISMODULE_CTX_FLAGS_LUA)) {
+      // Run query on Redis main thread.
+      context = CommandCtx_New(ctx, NULL, ast, argv[1], argv, argc);
+      context->tic[0] = tic[0];
+      context->tic[1] = tic[1];
+      _MGraph_Query(context);
+    } else {
+      // Run query on a dedicated thread.
+      RedisModuleBlockedClient *bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
+      context = CommandCtx_New(NULL, bc, ast, argv[1], argv, argc);
+      context->tic[0] = tic[0];
+      context->tic[1] = tic[1];
+      thpool_add_work(_thpool, _MGraph_Query, context);
+    }
 
     // Replicate only if query has potential to modify key space.
+    bool readonly = AST_ReadOnly(ast);
```

Comment:
Invalid read if we attached an AST to the QueryContext, since `ast` is freed by now.